content/reference/forms.md

```diff
@@ -0,0 +1,311 @@
+
+To validate the form's input fields, there's a few attributes we can use:
+- `pattern` - validate the form's input field using a Regular Expression. For
+  example `pattern="^.{1,15}$"` makes sure strings have a length of at least 1,
+  and not more than 15.
+- `required` - make sure that the field is filled in, and valid.
+- `title` - the message to display if the `pattern` attribute is invalid. This
+  is useful for users that can't read RegExes in their error messages (e.g.
+  everyone).
+
+Together these allow you to express a wide range of validation, and make sure
+your forms are filled in correctly and are accessible. Let's see what that
+that looks like:
+
+```html
+<form id="login" action="/dashboard">
+  <label for="username">
+    username
+  </label>
+  <input id="username" name="username"
+    type="text"
+    required
+    pattern=".{1,36}"
+    title="Username must be between 1 and 36 characters long."
+  >
+  <label for="password">
+    password
+  </label>
+  <input id="password" name="password"
+    type="password"
+    required
+  >
+  <input type="submit" value="Login">
+</form>
+```
+
+## Handling Form Submissions As Multipart
+So far we've seen how to create basic HTML forms with validation. This is a
+great starting point, but often we'll want to control submissions using
+javascript.
+
+Perhaps we can pre-populate some input fields. Perhaps there's input fields that
+rely on the values of other input fields. Starting off with JS from the start
+allows us to change the behavior without needing to change the architecture.
+
+Creating forms with Choo is almost identical to basic HTML. The main difference
+is that we create a `'submit'` event handler, and we control sending the data
+using `window.fetch()`.
+
+Let's create a form that sends data down as `'multipart/form-data`. We'll talk
+about how to submit it as JSON in the next section.
+
+```js
+var html = require('choo/html')
+var choo = require('choo')
+
+var app = choo()
+app.route('/', main)
+app.mount('body')
+
+function main () {                                                    // 1.
+  return html`
+    <body>
+      <form id="login" onsubmit=${onsubmit}>
+        <label for="username">
+          username
+        </label>
+        <input id="username" name="username"
+          type="text"
+          required
+          pattern=".{1,36}"
+          title="Username must be between 1 and 36 characters long."
+        >
+        <label for="password">
+          password
+        </label>
+        <input id="password" name="password"
+          type="password"
+          required
+        >
+        <input type="submit" value="Login">
+      </form>
+    </body>
+  `
+
+  function onsubmit (e) {                                              // 2.
+    e.preventDefault()                                                 // 3.
+    var form = document.querySelector('#login')                        // 4.
```

Comment:
I've been using Kory Nunn's [fastn.js](https://korynunn.github.io/fastn/) previously and trying to switch to Choo. Something that is nice in fastn that I feel like is missing here is that inputs have bindings.  Doing a document.querySelector('#login') doesn't feel right, it doesn't feel choo.

Is bindings a step too far?

**In the absence of bindings -** what about something like:

<form id="login" onsubmit=${onsubmit} onkeypress="${oninput}">
 <input value="${state.login.username}" >
...<><>

  function oninput (e){
      let inputValue = e.target.value + e.key;
      let name = e.srcElement.name;
      let value = {};
      value[name] = inputValue;
      emit('update:input','<formNamespace>', value);
    }

Reply:
This feedback made me dig in a little further. There's apparently `event.currentTarget` which gives us exactly what we need! https://developer.mozilla.org/en-US/docs/Web/API/Event/currentTarget

So yeah, no need for custom syntax :grin: :sparkles: